internal/vm/task.go

```diff
@@ -0,0 +1,338 @@
+	Task(string) (*Task, error)
+	TaskCount() uint
+	Remove(context.Context, string)
+	RemoveAll(context.Context)
+}
+
+// NewTaskManager initializes a new TaskManager
+func NewTaskManager() TaskManager {
+	return &taskManager{
+		tasks: make(map[string]*Task),
+	}
+}
+
+type taskManager struct {
+	mu    sync.RWMutex
+	tasks map[string]*Task
+}
+
+// AddTask registers a task with the provided metadata with the taskManager.
+// taskService should implement the TaskService API for the task (i.e. Create, Kill, Exec, etc.).
+func (m *taskManager) AddTask(
+	ctx context.Context,
+	containerID string,
+	taskService taskAPI.TaskService,
+	bundleDir bundle.Dir,
+	extraData *proto.ExtraData,
+	fifoSet *cio.FIFOSet,
+) (*Task, error) {
+	err := bundleDir.Create() // no-op if it already exists
+	if err != nil {
+		return nil, err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.tasks[containerID]; ok {
+		return nil, errors.Errorf("cannot add duplicate task with container ID %q", containerID)
+	}
+
+	task := &Task{
+		TaskService: taskService,
+		ID:          containerID,
+
+		extraData: extraData,
+		bundleDir: bundleDir,
+		fifoSet:   fifoSet,
+
+		cancelCh: make(chan struct{}),
+	}
+
+	m.tasks[task.ID] = task
+	return task, nil
+}
+
+func (m *taskManager) get(containerID string) (*Task, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	task, ok := m.tasks[containerID]
+	if !ok {
+		return nil, errors.Errorf("task with containerID %q not found ", containerID)
+	}
+
+	return task, nil
+}
+
+// assumes caller is holding m.mu lock
+func (m *taskManager) remove(ctx context.Context, containerID string, task *Task) {
+	delete(m.tasks, containerID)
+	err := task.CancelIO()
+	if err != nil {
+		log.G(ctx).WithField("id", containerID).Error("failed to cancel task io")
+	}
+}
+
+// Task returns the TaskService API for the task with the provided containerID, allowing
+// the caller to perform APIs like Create, Kill, Exec, etc. on the task.
+func (m *taskManager) Task(containerID string) (*Task, error) {
+	return m.get(containerID)
+}
+
+// TaskCount returns the number of tasks registered with the taskManager.
+func (m *taskManager) TaskCount() uint {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return uint(len(m.tasks))
+}
+
+// Remove unregisters the task with the provided containerID with the taskManager and cancels
+// any ongoing I/O proxying for the task.
+func (m *taskManager) Remove(ctx context.Context, containerID string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	task, ok := m.tasks[containerID]
+	if !ok {
+		// the task was already removed or just never existed, no-op
+		return
+	}
+
+	m.remove(ctx, containerID, task)
+}
+
+// RemoveAll calls Remove() on all tasks registered with the taskManager.
+func (m *taskManager) RemoveAll(ctx context.Context) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	for containerID, task := range m.tasks {
+		m.remove(ctx, containerID, task)
```

Comment:
How does this not deadlock? You are calling `m.mu.Lock` and then again in `Remove`

Reply:
This is calling `remove`, not `Remove`. `remove` is the internal-only method that doesn't acquire a lock

Reply:
Ah, missed the little r implementation.